File: src/main/java/AoC2019_11.java
import static java.util.stream.Collectors.toSet;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.OCR;
import com.github.pareronia.aoc.geometry.Draw;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.intcode.IntCode;
import com.github.pareronia.aoc.navigation.Heading;
import com.github.pareronia.aoc.navigation.NavigationWithHeading;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2019_11 extends AoCBase {
    
    private static final long BLACK = 0;
    private static final long WHITE = 1;
    private static final long LEFT = 0;
    private static final char FILL = '\u2592';
    private static final char EMPTY = ' ';
    
    private final List<Long> program;
    
    private AoC2019_11(final List<String> input, final boolean debug) {
        super(debug);
        assert input.size() == 1;
        this.program = IntCode.parse(input.get(0));
    }

    public static AoC2019_11 create(final List<String> input) {
        return new AoC2019_11(input, false);
    }

    public static AoC2019_11 createDebug(final List<String> input) {
        return new AoC2019_11(input, true);
    }

    private PaintJob paint(final long start) {
        final Set<Position> white = new HashSet<>();
        final NavigationWithHeading nav
            = new NavigationWithHeading(Position.of(0, 0), Heading.NORTH);
        final IntCode intCode = new IntCode(this.program, false);
        final Deque<Long> input = new ArrayDeque<>(List.of(start));
        final Deque<Long> output = new ArrayDeque<>();
        do {
            intCode.runTillInputRequired(input, output);
            if (output.pop() == WHITE) {
                white.add(nav.getPosition());
            } else {
                white.remove(nav.getPosition());
            }
            nav.turn(output.pop() == LEFT ? Turn.LEFT : Turn.RIGHT);
            nav.forward(1);
            input.add(white.contains(nav.getPosition()) ? 1L : 0L);
            intCode.runTillInputRequired(input, output);
        } while (!intCode.isHalted());
        final Set<Position> visited = nav.getVisitedPositions(false).stream()
                .distinct().collect(toSet());
        return new PaintJob(visited, white);
    }
    
    @Override
    public Integer solvePart1() {
        return paint(BLACK).visited.size();
    }
    
    @Override
    public String solvePart2() {
        final List<String> drawing = Draw.draw(paint(WHITE).white, FILL, EMPTY);
        drawing.forEach(this::log);
        return OCR.convert6(CharGrid.from(drawing), FILL, EMPTY);
    }

    public static void main(final String[] args) throws Exception {
        final Puzzle puzzle = Aocd.puzzle(2019, 11);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2019_11.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2019_11.createDebug(inputData)::solvePart2)
        );
    }
    
    record PaintJob(Set<Position> visited, Set<Position> white) { }
}

File: src/main/python/AoC2019_11.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 11
#

import aocd
from advent_of_code_ocr import convert_6

from aoc import my_aocd
from aoc.common import log
from aoc.geometry import Draw, Position, Turn
from aoc.intcode import IntCode
from aoc.navigation import Heading, NavigationWithHeading

BLACK = 0
WHITE = 1
LEFT = 0
FILL = "▒"
EMPTY = " "


def _parse(inputs: tuple[str, ...]) -> list[int]:
    return [int(_) for _ in inputs[0].split(",")]


def _paint(ints: list[int], start: int) -> tuple[set[Position], set[Position]]:
    white = set[Position]()
    nav = NavigationWithHeading(Position.of(0, 0), Heading.NORTH)
    int_code = IntCode(ints)
    input = [start]
    out: list[int] = []
    int_code.run_till_input_required(input, out)
    while True:
        if int_code.halted:
            break
        if out.pop(0) == WHITE:
            white.add(nav.position)
        else:
            white.discard(nav.position)
        nav.turn(Turn.LEFT if out.pop(0) == LEFT else Turn.RIGHT)
        nav.forward(1)
        input.append(1 if nav.position in white else 0)
        int_code.run_till_input_required(input, out)
    visited = {p for p in nav.get_visited_positions()}
    return white, visited


def part_1(inputs: tuple[str, ...]) -> int:
    _, visited = _paint(_parse(inputs), BLACK)
    return len(visited)


def part_2(inputs: tuple[str, ...]) -> str:
    white, _ = _paint(_parse(inputs), WHITE)
    drawing = Draw.draw(white, FILL, EMPTY)
    for _ in drawing:
        log(_)
    return convert_6(  # type:ignore[no-any-return]
        "\n".join(drawing), fill_pixel=FILL, empty_pixel=EMPTY
    )


def main() -> None:
    puzzle = aocd.models.Puzzle(2019, 11)
    my_aocd.print_header(puzzle.year, puzzle.day)

    inputs = my_aocd.get_input_data(puzzle, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()
